spock-core/src/main/java/org/spockframework/compiler/SpecParser.java

```diff
@@ -191,7 +192,10 @@ private void buildBlocks(Method method) throws InvalidSpecCompileException {
     Block currBlock = method.addBlock(new AnonymousBlock(method));
 
     for (Statement stat : stats) {
-      if (stat.getStatementLabel() == null)
+      String statementLabel = stat.getStatementLabel();
+      if ((statementLabel == null)
+        || (statementLabel.equals(DATA_TABLE_CROSS_PRODUCT_LABEL) && (currBlock instanceof WhereBlock)))
```

Comment:
Why the special handling, wouldn't it be cleaner to make it a proper block? Then the correct order could also be handled by `checkIsValidSuccessor`. Maybe treat it similar to `and:` so that it becomes part of the `where`.

Reply:
Hm, I didn't see it as block, as it is not a block on its own but just a combiner for two tables.
But the argument about `and:` is pretty strong of course. :-D